
<!DOCTYPE html>


<html lang="en" data-content_root="../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>mutations &#8212; Coding with Emily</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="../../_static/custom.css?v=cd1e198d" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../../_static/doctools.js?v=9bcbadda"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script src="../../_static/design-tabs.js?v=f930bc37"></script>
    <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'pages/problems/mutations';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="inheritance" href="inheritance.html" />
    <link rel="prev" title="genome" href="genome.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../_static/em_circle.png" class="logo__image only-light" alt=""/>
    <script>document.write(`<img src="../../_static/em_circle.png" class="logo__image only-dark" alt=""/>`);</script>
  
  
    <p class="title logo__title">Coding with Emily</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Learning</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../learning/start.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../learning/hello.html">Hello, world!</a></li>
<li class="toctree-l1"><a class="reference internal" href="../learning/variables.html">Variables</a></li>
<li class="toctree-l1"><a class="reference internal" href="../learning/conditions.html">Conditions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../learning/loops.html">Loops</a></li>
<li class="toctree-l1"><a class="reference internal" href="../learning/functions.html">Functions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../learning/imports.html">Imports</a></li>
<li class="toctree-l1"><a class="reference internal" href="../learning/files.html">Files</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Learning - Advanced</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../learning/random.html">Randomness</a></li>
<li class="toctree-l1"><a class="reference internal" href="../learning/vectorisation.html">Vectorisation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../learning/visualisation.html">Visualisation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Problems</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="greeting.html">greeting</a></li>
<li class="toctree-l1"><a class="reference internal" href="populations.html">populations</a></li>
<li class="toctree-l1"><a class="reference internal" href="genome.html">genome</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">mutations</a></li>
<li class="toctree-l1"><a class="reference internal" href="inheritance.html">inheritance</a></li>
<li class="toctree-l1"><a class="reference internal" href="survival.html">survival</a></li>
<li class="toctree-l1"><a class="reference internal" href="fastq.html">fastq</a></li>
<li class="toctree-l1"><a class="reference internal" href="scrabble.html">scrabble</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../_sources/pages/problems/mutations.rst" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.rst</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>mutations</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#concepts">Concepts</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#your-task">Your task</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#part-1-find-the-mutations">Part 1: Find the mutations</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#part-2-classify-the-mutations">Part 2: Classify the mutations</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#part-3-extra-credit-handle-frame-shift-mutations">Part 3: (Extra credit) Handle frame-shift mutations</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#part-3a-implement-needleman-wunsch">Part 3a: Implement Needleman-Wunsch</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="mutations">
<h1>mutations<a class="headerlink" href="#mutations" title="Link to this heading">#</a></h1>
<figure class="align-center" id="id1">
<a class="reference internal image-reference" href="../../_images/mutation_breakup.png"><img alt="Two DNA strands breaking up" src="../../_images/mutation_breakup.png" style="width: 500px;" />
</a>
<figcaption>
<p><span class="caption-text">It’s not you, it’s not me, it’s just mutations.</span><a class="headerlink" href="#id1" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p>In <a class="reference external" href="genome.html">the previous problem</a>, we learned about genomes, nucleotide sequences, and how ribosomes
build proteins from those sequences. In this problem, we’ll consider how mutations can cause things to go
wrong when trying to build those proteins.</p>
<section id="concepts">
<h2>Concepts<a class="headerlink" href="#concepts" title="Link to this heading">#</a></h2>
<p>First, let’s review some key concepts (that I may or may not have only just learned myself!). We’re can think
about two different types of mutations that can occur in a nucleotide sequence:</p>
<ul class="simple">
<li><p><strong>Point mutations</strong>: A single nucleotide is changed to a different nucleotide. For example, an “A” might be changed to a “G”.</p></li>
<li><p><strong>Frame-shift mutations</strong>: One or more nucleotides are added to or removed from the sequence. For example, an “A” might be inserted into the sequence, or a “T” might be deleted.</p></li>
</ul>
<figure class="align-center" id="id2">
<a class="reference internal image-reference" href="../../_images/point_vs_frame_shift.png"><img alt="Point mutation vs frame-shift mutation" src="../../_images/point_vs_frame_shift.png" style="width: 600px;" />
</a>
<figcaption>
<p><span class="caption-text">Some mutations have bigger effects than others.</span><a class="headerlink" href="#id2" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p>You can imagine that a point mutation will have a smaller effect on the resulting protein than a frame-shift
mutation, since the latter will change the reading frame of all subsequent codons.</p>
<p>We can further classify mutations based on their effect on the resulting protein:</p>
<ul class="simple">
<li><p><strong>Silent mutations</strong>: The mutation does not change the amino acid that is produced. For example, both “GAA”and “GAG” code for the amino acid glutamic acid, so a mutation from “GAA” to “GAG” would be silent.</p></li>
<li><p><strong>Missense mutations</strong>: The mutation changes the amino acid that is produced. For example, a mutation from “GAA” to “GAC” would change the amino acid from glutamic acid to aspartic acid.</p></li>
<li><p><strong>Nonsense mutations</strong>: The mutation changes a codon to a stop codon, causing the protein to be truncated. For example, a mutation from “GAA” to “TAA” would change the codon to a stop codon.</p></li>
</ul>
<div class="admonition-required-humour admonition">
<p class="admonition-title">Required humour</p>
<p>You should now check your learning by looking at the first image again now you know what a nonsense mutation is! If you don’t laugh, you must not have understood the concept properly, there’s no other explanation.</p>
</div>
<p>For the purposes of this problem, we’ll only be thinking about point mutations, but if you’re keen there’s an extra credit section at the end where you can try to handle frame-shift mutations as well.</p>
</section>
<section id="your-task">
<h2>Your task<a class="headerlink" href="#your-task" title="Link to this heading">#</a></h2>
<p>Imagine you have a reference nucleotide sequence (the “correct” sequence) from some perfect organism that has no mutations. Next you have a second nucleotide sequence from an organism that may have some mutations relative to the reference sequence.</p>
</section>
<section id="part-1-find-the-mutations">
<h2>Part 1: Find the mutations<a class="headerlink" href="#part-1-find-the-mutations" title="Link to this heading">#</a></h2>
<p>First, implement <code class="docutils literal notranslate"><span class="pre">find_point_mutations(reference,</span> <span class="pre">mutated)</span></code> which takes in two nucleotide sequences (strings) and returns a list of locations (0-indexed) where the point mutations occur. You can assume that both sequences are the same length and only differ by point mutations (i.e., no insertions or deletions). You can also assume that the sequences only contain the characters “A”, “C”, “G”, and “T”.</p>
<p>Here are some test cases to get you started:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">find_point_mutations</span><span class="p">(</span><span class="s2">&quot;ACGT&quot;</span><span class="p">,</span> <span class="s2">&quot;ACGT&quot;</span><span class="p">)</span>  <span class="c1"># should return []</span>
<span class="n">find_point_mutations</span><span class="p">(</span><span class="s2">&quot;ACGT&quot;</span><span class="p">,</span> <span class="s2">&quot;AGGT&quot;</span><span class="p">)</span>  <span class="c1"># should return [1]</span>
<span class="n">find_point_mutations</span><span class="p">(</span><span class="s2">&quot;ACGT&quot;</span><span class="p">,</span> <span class="s2">&quot;TGCA&quot;</span><span class="p">)</span>  <span class="c1"># should return [0, 1, 2, 3]</span>
</pre></div>
</div>
</section>
<section id="part-2-classify-the-mutations">
<h2>Part 2: Classify the mutations<a class="headerlink" href="#part-2-classify-the-mutations" title="Link to this heading">#</a></h2>
<p>Now the real question is: what effect do those mutations have on the resulting protein? To answer this, implement <code class="docutils literal notranslate"><span class="pre">classify_mutations(reference,</span> <span class="pre">mutated)</span></code> which takes in two nucleotide sequences (strings) and returns a list of tuples, where each tuple contains the index of the mutation and its classification (“silent”, “missense”, or “nonsense”).</p>
<p>You can use the genetic code table from the previous problem to determine the amino acids produced by each codon.</p>
<p><strong>To simplify things</strong>, the sequences that you receive will always start with a start codon (“ATG”) and end with a stop codon (“TAA”, “TAG”, or “TGA”), and the length of the sequences will always be a multiple of 3. There will also only be point mutations (no insertions or deletions), and only one mutation will occur per codon.</p>
<p>Recall that silent mutations do not change the amino acid, missense mutations change the amino acid, and nonsense mutations change a codon to a stop codon.</p>
<p>Here are some test cases to get you started:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">reference</span> <span class="o">=</span> <span class="s2">&quot;ATGGAGCCATAA&quot;</span>

<span class="n">classify_mutations</span><span class="p">(</span><span class="n">reference</span><span class="p">,</span> <span class="s2">&quot;ATGGAGCCATAA&quot;</span><span class="p">)</span>   <span class="c1"># should return []</span>
<span class="n">classify_mutations</span><span class="p">(</span><span class="n">reference</span><span class="p">,</span> <span class="s2">&quot;AGGGAGCCGTAA&quot;</span><span class="p">)</span>   <span class="c1"># should return [(1, &quot;missense&quot;)]</span>
<span class="n">classify_mutations</span><span class="p">(</span><span class="n">reference</span><span class="p">,</span> <span class="s2">&quot;ATGGAGCCCTAA&quot;</span><span class="p">)</span>   <span class="c1"># should return [(8, &quot;silent&quot;)]</span>
<span class="n">classify_mutations</span><span class="p">(</span><span class="n">reference</span><span class="p">,</span> <span class="s2">&quot;ATGTAGCCCTAA&quot;</span><span class="p">)</span>   <span class="c1"># should return [(3, &quot;nonsense&quot;), (8, &quot;silent&quot;)]</span>
</pre></div>
</div>
</section>
<section id="part-3-extra-credit-handle-frame-shift-mutations">
<h2>Part 3: (Extra credit) Handle frame-shift mutations<a class="headerlink" href="#part-3-extra-credit-handle-frame-shift-mutations" title="Link to this heading">#</a></h2>
<p>A much more difficult problem is to handle frame-shift mutations, where nucleotides are inserted or deleted from the sequence. This changes the reading frame of all subsequent codons, which can have a drastic effect on the resulting protein.</p>
<p>To identify all insertions, deletions, and substitutions between two sequences via brute force you would need to try all possible ways of aligning the two sequences, which is computationally expensive. We can calculate exactly that cost for two sequences of length $n$ and $m$, since the number of possible alignments is given by</p>
<div class="math notranslate nohighlight">
\[\text{Number of alignments} = \sum_{k=0}^{\min(n, m)} \binom{m}{k} \binom{n}{k} 2^k\]</div>
<p>Which means you algorithm time complexity is exponential, approximately</p>
<div class="math notranslate nohighlight">
\[T_{\rm brute} = O\left( 2^{n \cdot m} \right)\]</div>
<p>That’s clearly computationally infeasible for even moderately sized sequences! But fear not, there is a well-known dynamic programming algorithm called the <a class="reference external" href="https://en.wikipedia.org/wiki/Needleman%E2%80%93Wunsch_algorithm">Needleman-Wunsch algorithm</a> that can find the optimal alignment in polynomial time, specifically</p>
<div class="math notranslate nohighlight">
\[T_{\rm NW} = O(n \cdot m)\]</div>
<p>We’re going to work together here to implement this algorithm to align two sequences, and then use that alignment to identify mutations.</p>
<section id="part-3a-implement-needleman-wunsch">
<h3>Part 3a: Implement Needleman-Wunsch<a class="headerlink" href="#part-3a-implement-needleman-wunsch" title="Link to this heading">#</a></h3>
<p>First, we’re going to implement the Needleman-Wunsch algorithm itself. There’s a function called <code class="docutils literal notranslate"><span class="pre">needleman_wunsch(s1,</span> <span class="pre">s2,</span> <span class="pre">match=1,</span> <span class="pre">mismatch=-1,</span> <span class="pre">gap=-2)</span></code> that takes in two sequences (strings) and returns a tuple of two aligned sequences (strings), where gaps are represented by the “-” character. The parameters <code class="docutils literal notranslate"><span class="pre">match</span></code>, <code class="docutils literal notranslate"><span class="pre">mismatch</span></code>, and <code class="docutils literal notranslate"><span class="pre">gap</span></code> are the scores for matching characters, mismatched characters, and gaps, respectively.</p>
<p>Your function should do the following steps:</p>
<ol class="arabic simple">
<li><p>Create a scoring matrix that has size n x m, where n and m are the lengths of the two sequences.</p></li>
<li><p>Initialize the first row of the matrix, each cell is equal to the one to its left plus the gap penalty.</p></li>
<li><p>Initialize the first column of the matrix, each cell is equal to the one above it plus the gap penalty.</p></li>
<li><dl class="simple">
<dt>Fill in the rest of the matrix using the following rules:</dt><dd><ul class="simple">
<li><p>Calculate three scores, one for a diagonal move (match or mismatch), one for an up move (gap in sequence 2), and one for a left move (gap in sequence 1).</p></li>
<li><p>The diagonal score is the score from the cell diagonally above and to the left plus either the match score (if the characters are the same) or the mismatch score (if they are different).</p></li>
<li><p>The up score is the score from the cell above plus the gap penalty.</p></li>
<li><p>The left score is the score from the cell to the left plus the gap penalty.</p></li>
<li><p>The score for the current cell is the maximum of the three scores.</p></li>
<li><p>Based on which score was chosen, record the move (diagonal, up, or left) in a traceback matrix (as a ‘D’, ‘U’, or ‘L’).</p></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>Once the matrix is filled in, perform a traceback starting from the bottom-right cell to construct the aligned sequences.</dt><dd><ul class="simple">
<li><p>If the move is diagonal, add the corresponding characters from both sequences to the aligned sequences.</p></li>
<li><p>If the move is up, add the character from sequence 1 and a gap to sequence 2.</p></li>
<li><p>If the move is left, add a gap to sequence 1 and the character from sequence 2.</p></li>
</ul>
</dd>
</dl>
</li>
<li><p>Reverse the aligned sequences since they were constructed backwards during the traceback.</p></li>
<li><p>Return the aligned sequences as a tuple.</p></li>
</ol>
</section>
</section>
</section>


                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="genome.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">genome</p>
      </div>
    </a>
    <a class="right-next"
       href="inheritance.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">inheritance</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#concepts">Concepts</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#your-task">Your task</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#part-1-find-the-mutations">Part 1: Find the mutations</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#part-2-classify-the-mutations">Part 2: Classify the mutations</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#part-3-extra-credit-handle-frame-shift-mutations">Part 3: (Extra credit) Handle frame-shift mutations</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#part-3a-implement-needleman-wunsch">Part 3a: Implement Needleman-Wunsch</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Tom Wagg
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2025, Tom Wagg.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>